### PR TITLE
Distributions and Environments integration

### PR DESCRIPTION
**Pull request by KanuX-14**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -46,19 +46,26 @@ You have multiple options:
 
 - Basic building tools
 - CMake
-- Qt5, QtWebEngine and Kf5Notifications
+- Qt5 and QtWebEngine
 - **PipeWire** (it currently doesn't work with PulseAudio)
 - Git
+- notification-daemon
```
Comment (maltejur) [inline README.md:52]: Is this a dependency which has to be installed?

File: README.md [hunk 2]
```diff
@@ -46,19 +46,26 @@ You have multiple options:
...
+- Kf5Notifications (optional, for KDE users)
```
Comment (maltejur) [inline README.md:53]:
```suggestion
- Kf5Notifications (optional, for better desktop notifications)
```

**Review by maltejur — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (KanuX-14) [reply README.md:52]: > Is this a dependency which has to be installed?

No, this is the basic notification daemon. KDE is usually bloated and was a pain to remove knotifications from my system.
What about making a installation script to identify the DEs?

Comment (KanuX-14) [reply README.md:52]: Various [debian packages](https://packages.debian.org/buster/notification-daemon), for example, use notification-daemon as part of the recipe.

**Comment by KanuX-14** [comment]
I might implement a different notification system, setting it optional. I will tell when the repository is ready to merge.

**Comment by maltejur** [comment]
> I might implement a different notification system, setting it optional. I will tell when the repository is ready to merge.

What do you mean exactly? There are already two ways for a noticiation to show, why a third?

**Comment by maltejur** [comment]
Or are you talking about automatically detecting if kf5notifications is available? I have that already implemented on [this branch](https://github.com/maltejur/discord-screenaudio/tree/keybinds], but there are also some other changes there which arent ready yet.

**Comment by KanuX-14** [comment]
> What do you mean exactly? There are already two ways for a noticiation to show, why a third?

I mean, it's so small. Changing the font size would help my eyes haha
Also thought about using `notify-send` through the system, but would have to fetch the name, icon, and the text... maybe just making the current Qt notification better readable.

**Comment by maltejur** [comment]
Wait, so the standard qt notification isn't a normal system notification for you? Interesting, I am on KDE so it is completely integrated there.

**Comment by KanuX-14** [comment]
> Wait, so the standard qt notification isn't a normal system notification for you? Interesting, I am on KDE so it is completely integrated there.

No. The standard notification here is `notification-daemon`. Without it, I won't receive any notifications at all.

**Comment by KanuX-14** [comment]
One thing that I couldn't make is to show the icon. Do you have any idea @maltejur ?

**Comment by maltejur** [comment]
I am currently not home so sorry for not working on your pull request so far. I am back in a week.

**Comment by KanuX-14** [comment]
No problem. There are few touches here and there that I need to make it look acceptable.

File: README.md [hunk 3]
```diff
@@ -46,19 +46,50 @@ You have multiple options:
 
 - Basic building tools
 - CMake
-- Qt5, QtWebEngine and Kf5Notifications
+- Qt5 and QtWebEngine
 - **PipeWire** (it currently doesn't work with PulseAudio)
 - Git
+- notification-daemon
```
Comment (maltejur) [inline README.md:52]:
```suggestion
- notification-daemon (optional, for basic desktop notifications)
```

**Comment by maltejur** [comment]
> One thing that I couldn't make is to show the icon. Do you have any idea @maltejur ?

This works for me:

`notify-send --icon de.shorsh.discord-screenaudio Hello`

image

File: src/mainwindow.cpp [hunk 4]
```diff
@@ -51,6 +59,19 @@ void MainWindow::setupWebView() {
...
+      QByteArray message_ba = notificationInfo->message().toLocal8Bit();
+      char image_path[1024], command[2048];
+      const char *title = title_ba.data();
+      const char *message = secureMe(message_ba.data());
+      snprintf(image_path, 32+strlen(title), "/tmp/discord-screenaudio-%s.png", title);
+      notificationInfo->icon().save(image_path);
+      snprintf(command, 70+strlen(image_path)+strlen(title)+strlen(message), "notify-send --icon=\"%s\" --app-name=\"discord-screenaudio\" \"%s\" \"%s\"", image_path, title, message);
+      system(command);
+    });
 #endif
```
Comment (maltejur) [inline src/mainwindow.cpp:75]: You could move the `#ifdef KNOTIFICATIONS` into the lambda to make it a bit less redundant.

**Comment by KanuX-14** [comment]
> > One thing that I couldn't make is to show the icon. Do you have any idea @maltejur ?
> 
> This works for me:
> 
> `notify-send --icon de.shorsh.discord-screenaudio Hello`
> 
> image

I mean, for the users' images. I made it by saving the image to /tmp folder. This one can apply by the installation process.

**Comment by KanuX-14** [comment]
Review the code, if you find any errors or changes you want to make.
I have tried to do a "Reply" button. Successfully could make it interact with the application, but it opened another one for some reason.

File: src/mainwindow.cpp (hunk 4 above)
Comment (KanuX-14) [reply src/mainwindow.cpp:75]: If you know how, do it. I don't know what lambda is, to be honest.

**Comment by maltejur** [comment]
> I mean, for the users' images. I made it by saving the image to /tmp folder. This one can apply by the installation process.

Oh yeah I see you already solved it. But I am having a few problems with it, one of those errors always occurs when a notification should be shown:

Screenshot_20220829_193240
Screenshot_20220829_193321

However when I add a `qDebug() << command;` in there, just for checking the command, it works just fine again. Some weird C++ thing is going on here.

I think it would be best if you would just rewrite that code using Qt stuff, to fix the error and to generally make the code more consistent.

**Comment by KanuX-14** [comment]
> Some weird C++ thing is going on here.
Maybe because I used a function in C to return a "safe" message value back to the command. I tried doing the filter in C++, but it didn't let me.

**Comment by KanuX-14** [comment]
I can guess what it could be. Some notifications don't show up sometimes, maybe because of the string getting a value out of memory or weird value return.

File: install.sh [hunk 5]
```diff
@@ -0,0 +1,105 @@
+#!/bin/sh
```
Comment (maltejur) [inline install.sh:1]: Some stuff about the install script:

- There already exists a very simple script in `scripts/install.sh`, you should probably replace that an move the file there.
- Why do you set the `$CURRENT_HOME` to `$HOME` first and then have it overridden again? Can't you just use `$HOME`?
- I dislike that the home directory is polluted by this script, maybe something like `~/.local/share/discord-screenaudio/git` would be a better place for the repo.

Comment (KanuX-14) [reply install.sh:1]: > Why do you set the $CURRENT_HOME to $HOME first and then have it overridden again?

Sorry. I copied it from a old script. Thank you for seeing it.

> There already exists a very simple script in scripts/install.sh

I thought of replacing it, but didn't wanted to do that without permission.

> I dislike that the home directory is polluted by this script

True. I am already doing the changes for the installation script.

File: src/mainwindow.cpp (hunk 4 above)
Comment (maltejur) [reply src/mainwindow.cpp]: I am just talking about the callback function passed to `setNotificationPresenter`. I've pushed the required changes.

**Comment by maltejur** [comment]
> I think it would be best if you would just rewrite that code using Qt stuff, to fix the error and to generally make the code more consistent.

See https://github.com/KanuX-14/discord-screenaudio/pull/1

File: scripts/install.sh [hunk 6]
```diff
@@ -1,6 +1,104 @@
...
+		"pacman")
+			if [ -z "$(pacman -Q | grep "$package")" ]; then
+				sudo pacman -S --noconfirm $package
+				printf "$GREEN""$PURPLE""$package""$GREEN"" has been installed!""$RESET_COLOUR""\n"
+			else
+				printf "$PURPLE""$package""$RESET_COLOUR"" has already been satisfied!\n"
+			fi
+			;;
+	esac
+done
```
Comment (maltejur) [inline scripts/install.sh:96]:
```suggestion

# Install the packages.
case $package_manager in
	"apt")
		sudo apt-get install -y build-essential git cmake qtbase5-dev qtwebengine5-dev pkg-config libpipewire-0.3-dev $de_packages
		;;
	"pacman")
		sudo pacman -S --noconfirm git cmake qt5-base qt5-webengine xdg-desktop-portal pipewire $de_packages
		;;
esac
```

File: scripts/install.sh [hunk 7]
```diff
@@ -1,6 +1,104 @@
...
+PURPLE="\e[1;35m"
+CYAN="\e[1;36m"
+RESET_COLOUR="\e[0m"
+
+# Set home based on user or root
+if [ ${CURRENT_USER} == "root" ]; then
+	CURRENT_HOME="/root"
+else
+	CURRENT_HOME="/home/${CURRENT_USER}"
+fi
```
Comment (maltejur) [inline scripts/install.sh:24]: As I said earlier, I would find it better if the script doesn't clone the repo to the home directory

**Review by maltejur — CHANGES_REQUESTED** [review]

File: scripts/install.sh (hunk 7 above)
Comment (KanuX-14) [reply scripts/install.sh:24]: Sorry, completely forgot it.

File: scripts/install.sh (hunk 6 above)
Comment (KanuX-14) [reply scripts/install.sh:96]: Do you have a better idea on how to check the packages without requiring root privileges?
I made this check to run straight to the point. Saves some time after each compile.

**Comment by federicotorrielli** [comment]
Regarding the notification system: distributions like Pop_os don't use libnotify but use the standard systemd-notify to deliver notifications, keep that in mind :)

**Comment by KanuX-14** [comment]
> libnotify but use the standard systemd-notify to deliver notifications

Thank you about this fact. I personally don't use Pop!_OS. If you know a way to implement it, feel free to push.

**Comment by federicotorrielli** [comment]
I will look into this in this week or so. This doesn't only apply to Pop, there are many distributions using systemd-notify, ofc

File: scripts/install.sh (hunk 7 above)
Comment (KanuX-14) [reply scripts/install.sh]: Fixed by recent commit.

**Review by KanuX-14 — COMMENTED** [review]
All ok.

**Comment by KanuX-14** [comment]
Closing, as it was already implemented.

**Pull request closed without merging**